### ISSUE TITLE
Initial Models with ROC curves

### ISSUE THREAD

**Issue by poltergeistjoya** [poltergeistjoya/Vacant_Lot_Detection#2]
# Modeling Plan: Vacant Lot Detection

## Context

EDA confirmed spectral signals (SAVI,BareSoilProxy, NDVI, Brightness) in NAIP imagery that distinguish vacant lots from occupied parcels. The immediate goal is a parcel-level classifier that validates whether those signals are strong enough for supervised classification. 

Segmentation (pixel-level inference without parcel boundaries) is the long-term goal but is **deferred** until parcel-level performance is established.

---

## Design Decisions

- **No geometric features.** Shape_Area, perimeter, etc. will not be available at inference time on cities without parcel data. They must not appear in any model even during this parcel-level stage.
- **Labels**: Use `CityConfig.parcel.vacant_codes` applied to `CityConfig.parcel.landuse_column` (BldgClss in Mappluto G7, and V*).
- **No building-footprint heuristic** (NumBldgs=0 etc.) as a label source — too NYC-specific.

> **TODO (future)**: Compare ROC curves between `LandUse only` vs `LandUse + building heuristic` label strategies to quantify the signal the heuristic adds.

---

## Features

**8 NAIP spectral bands**: R, G, B, N (NIR), NDVI, SAVI, Brightness, BareSoilProxy.

Each band is aggregated over all pixels within a parcel → **mean** and **stdDev** per band = **16 model inputs**.

- `mean` — average spectral signature of the parcel
- `stdDev` — internal texture/heterogeneity of the parcel; useful proxy for simpler models and consistent with what future pixel-level models will have access to

Count columns are excluded — pixel count is a proxy for parcel area, which is geometric information that won't be available at inference time on cities without parcel data.

---

## Naming Convention

`{city}_{naip_year}_s{stage}_{model_type}_{run_id}`

Example: `nyc_2022_s1_rf_001`

Each saved model gets a sidecar `.json` with the same base name storing hyperparameters, feature columns, label strategy, and CV results.

---

## Train / Test Split (NOT CONFIRMED)

**Spatial holdout by borough** — random splits are too optimistic due to spatial autocorrelation between neighboring parcels.

- **Train**: Manhattan, Brooklyn, Queens
- **Test**: Bronx, Staten Island

Multi-year temporal splits (different NAIP years) are a future extension requiring a separate GEE export.

---

### New files
- `EDA/stage1_parcel_classifier.ipynb`
- `modeling.py`

### Label construction (`modeling.py`)

```python
def build_labels(df: pd.DataFrame, cfg: CityConfig) -> pd.Series:
    """Binary vacancy label using CityConfig vacant_codes on landuse_column."""
    return df[cfg.parcel.landuse_column].isin(cfg.parcel.vacant_codes).astype(int)
```

### Class imbalance (~8% positive rate)

sklearn does NOT handle imbalance by default — must be set explicitly:
- Logistic Regression / SVM: `class_weight="balanced"` in the constructor
- Random Forest: `class_weight="balanced_subsample"` in the constructor

sklearn computes the actual per-class weights from training data frequency automatically once the parameter is set.

**Evaluation**: F1 and Average Precision (area under precision-recall curve). Do not report accuracy. Inspect the full precision-recall curve visually — threshold tuning to a specific recall target is an operational decision deferred until the model performs reasonably.

### Models

1. **Logistic Regression** — linear baseline; coefficient magnitudes show which features matter
2. **SVM (RBF kernel)** — non-linear boundary via `K(x,y) = exp(-γ||x−y||²)`
3. **Random Forest** — non-linear, handles feature interactions, feature importance rankings

Evaluation: Stratified 5-fold CV on the train set. Report mean ± std for F1 and Average Precision. Borough holdout (Bronx + Staten Island) used as final test set.

### `modeling.py` utilities

```python
build_labels(df, cfg) -> pd.Series
get_feature_matrix(df, feature_cols, label_col) -> (X, y)
evaluate_classifier(model, X_val, y_val) -> dict   # F1, Average Precision, PR curve data
save_model(model, path, metadata: dict)             # saves .joblib + sidecar .json
load_model(path) -> (model, metadata)
```

### Outputs

```
EDA/outputs/models/nyc_2022_s1_logreg_001.joblib + .json
EDA/outputs/models/nyc_2022_s1_svm_001.joblib    + .json
EDA/outputs/models/nyc_2022_s1_rf_001.joblib     + .json
EDA/outputs/figures/nyc_2022_s1_pr_curves.png
EDA/outputs/figures/nyc_2022_s1_feature_importance.png
EDA/outputs/data/nyc_2022_s1_cv_results.csv
```

---

## Future Directions

**Stage 2 — Pixel-level classifier**: Train a sliding-window RF on per-pixel spectral windows extracted from the parcel raster + NAIP. Deferred because the appropriate window size is non-trivial (minimum lot is ~50 m² = 50 pixels ≈ 7×7, so any window larger than ~5×5 bleeds into neighboring parcels for small lots).

**Stage 3 — Semantic segmentation (U-Net)**: Export NAIP as raw tiles, generate pixel-level label masks from parcel classifications, train a U-Net (ResNet-34 encoder, 4-channel NAIP input) for fully parcel-free inference. Requires GPU. The parcel raster and label masks from Stage 1 become the training data.

---

## Critical Existing Files

| File | Role |
|------|------|
| `config.py` | `CityConfig.parcel.landuse_column` + `vacant_codes` — label abstraction |
| `mappluto.py` | `load_and_sample()` produces merged spectral + MapPLUTO DataFrame |
| `analysis.py` | Scaler and feature prep conventions to follow in `modeling.py` |
| `gee_utils.py` | `read_csv_from_gcs()` loads spectral stats |
| `EDA/clean.ipynb` | Source of merged spectral + metadata DataFrame |
